### PR TITLE
[WIP] Proposal for a new configuration format

### PR DESCRIPTION
**Pull request by ambitious-octopus**
# Problem
We have many parameters that include different modes of operation in a single file.  The solution we are currently using is a single `default.yaml ` file that is used both as a parent file where the default parameters are taken and as an option to change these parameters. The problem is that this file is inconvenient for both our code and our users. It does not contain all the information we want and also it is quite confusing for the user who does not know that many parameters here are used for different modes in different ways.

# Solution
We massively extend the file. For each mode we put all the parameters needed by both humans and computers.
1. We divide the file into modes, (train, val, augmentation, etc.)
2. For each single parameter we specify all the things that can be useful both for documentation and for testing and for operation. I have put some ideas at the moment but we can extend. For instance, for each parameter, I have specified the typing of that parameter. I have put a description to make it readable to humans, this can be useful to show more interesting errors or build documentation tables or direct users to the correct places. Each parameter has a minimum and a maximum, if the user goes out of range we can throw a warning. Or we can develop tests that cover the entire range. Of course, not all parameters fit a linear distribution, for example, for the learning rate we can specify a logarithmic scale to test. 

This change is massive I know, that's why I want to hear your opinion. @glenn-jocher @lakshanthad @fcakyon @Laughing-q @RizwanMunawar @Burhan-Q @IvorZhu331 

## Extra
yaml is very slow! I also added a test with a JSON file, less human readable but very fast. Since we'll be calling this file a few times, we like a little speed.

```python
import yaml
import time
import json

start = time.time()
with open("ultralytics/cfg/default.yaml", errors="ignore", encoding="utf-8") as f:
    s = f.read()
    data = yaml.safe_load(s) or {}

print(f"Time taken to load yaml file: {time.time() - start:.4f} seconds")

start = time.time()
with open("ultralytics/cfg/default.json", "r") as f:
    data = json.load(f) 

print(f"Time taken to load JSON file: {time.time() - start:.4f} seconds")
```

## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
A new default configuration JSON file is introduced to define settings for model training, export, hyperparameters, and data augmentation, aligning with YAML settings.

### 📊 Key Changes
- **New Configuration File**: Added `default.json`, providing a structured configuration with descriptions for settings.
- **Detailed Export Options**: Defined multiple export formats such as ONNX, TorchScript, and more, with detailed parameters for optimization and device compatibility.
- **Hyperparameter Definitions**: Added descriptions and default values for key training parameters, including learning rates, momentum, and regularization.
- **Augmentation Techniques**: Detailed parameters for various image augmentations like translation, scaling, and hue adjustments.
  
### 🎯 Purpose & Impact
- **Consistency & Ease of Use**: Simplifies model setup by providing a clear and editable configuration template, enhancing user understanding of each parameter's role.
- **Improved Portability**: Supports multiple export formats, enabling users to deploy models across various platforms and devices easily.
- **Enhanced Training Flexibility**: Offers fine-grained control over training strategies and augmentations, leading to potentially better model performance and generalization.

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @ambitious-octopus, thank you for submitting a pull request to the `ultralytics/ultralytics` repository! 🚀 This is an automated response to assist you in ensuring a smooth integration of your proposed changes.

Here's a checklist to help streamline the review process:

- ✅ **Define a Purpose**: Make sure your PR description clearly explains the purpose of your proposed configuration format and its benefits for both code and users. Link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues) if applicable. Ensure your commit messages are clear and aligned with our project's conventions.
- ✅ **Synchronize with Source**: Please confirm that your PR is up to date with the `main` branch of `ultralytics/ultralytics`. If it's behind, you can update it by clicking the 'Update branch' button or by executing `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify that all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are green. If any checks fail, kindly address the issues to ensure smooth integration.
- ✅ **Update Documentation**: If your changes affect documentation, please update the relevant [documentation](https://docs.ultralytics.com) sections.
- ✅ **Add Tests**: Where applicable, include or update tests to cover your changes. Ensure all tests pass successfully.
- ✅ **Sign the CLA**: If you haven't done so already, please sign our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) by adding "I have read the CLA Document and I sign the CLA" in a comment if this is your first Ultralytics contribution.
- ✅ **Minimize Changes**: Aim to limit changes to the **essential** ones needed for your feature or fix. 💡 _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

For further information, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing). An Ultralytics engineer will review your PR soon to provide additional feedback. If you have any questions, feel free to leave a comment. Thank you for contributing to Ultralytics! 🌟

**Comment by Burhan-Q** [comment]
🤣 @ambitious-octopus great minds think like me `/s` check out #10163

**Comment by Burhan-Q** [comment]
I agree, but Glenn mentioned that unifying things in this way would make it much more difficult to read or find the config information (at least in my implementation). Also, if you recall, after introducing #14706 there was discussion about completely removing the config YAML file, which I still think is the right direction to go, but that's just my opinion

**Comment by glenn-jocher** [comment]
@Burhan-Q thank you for sharing your thoughts! The ideas around unifying or even reconsidering the use of YAML configs have sparked valuable discussions. While opinions differ, any changes need to balance usability, performance, and clarity. If you'd like to contribute further, feel free to propose enhancements or alternatives that address these concerns directly.
